KEP-5018: move to beta in 1.34 (#5327)
kubernetes/enhancements: keps/sig-auth/5018-dra-adminaccess/README.md

Reviewer: while reviewing an unrelated DRA KEP, we realized this is using an inconsistent label key ... 

labels / annotations are expected to use `$group.kubernetes.io/...` domain prefixes, so this should really be `resource.kubernetes.io/admin-access`

/hold

Author: so do we need to update this label in 1.34? and if so, is that still ok to move to beta?

Reviewer: > so do we need to update this label in 1.34?

I would say so, yes.

> and if so, is that still ok to move to beta?

I think so, for the following reasons:
1. pre-1.34 clusters default the feature off, so the adminAccess field is forcibly cleared on creation, so there's no existing objects with this field set unless someone opted into an alpha feature
2. ResourceClaim / ResourceClaimTemplate only check admin access on create, since their specs are immutable, so no 1.33 server would have to handle authorizing an update to an `adminAccess: true` object persisted by a 1.34 server
3. If someone was using the feature in alpha, it's easy to double-label their namespace until they complete upgrading to 1.34 to let both 1.33 and 1.34 be happy.

Author: Thanks for catching this now! and +1 on moving to beta. 
I can add a validation and warning for using the old label, not sure if it's worth the effort to maintain the code though.

Reviewer: > Thanks for catching this now! and +1 on moving to beta. I can add a validation and warning for using the old label, not sure if it's worth the effort to maintain the code though.

I think just a release note is fine for something changed before exiting alpha.

   ```yaml
   metadata:
     labels:
       resource.kubernetes.io/admin-access: "true"
   ```

>    Note: This label has been updated from `resource.k8s.io/admin-access` while the feature was in alpha in v1.33.

Reviewer 2: This is not quite accurate, b/c 1.33 is still using `resource.k8s.io/admin-access: "true"`, so maybe
```suggestion
   Note: This label has been updated from `resource.k8s.io/admin-access` before the beta promotion.
```
or
```suggestion
   Note: This label has been updated from `resource.k8s.io/admin-access` while the feature was in alpha.
```
Ideally I'd say open a PR doing so asap.

> - Gather feedback

Reviewer 2: Missing bits higher in the doc:
1. make sure to check appropriate boxes in `Release Singoff Checklist`
2. In `Integration tests` section, please make sure to link tests according [to the template](https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md#integration-tests), especially the newly added that are called out there, since looking at the PRs submitted during alpha they did add new tests.

Reviewer 2: These comments still hold.

- Additional tests are in Testgrid and linked in KEP

> - Implementations in the kubernetes-sigs/dra-example-driver: https://github.com/kubernetes-sigs/dra-example-driver/issues/97 and the NVIDIA dra driver: https://github.com/NVIDIA/k8s-dra-driver-gpu/issues/337

Reviewer: do those issues mean we will show those repos labeling namespaces as admin access and using devices as admin access before promoting the gate to beta?

Author: Yes, `Implementations in the kubernetes-sigs/dra-example-driver` was part of the original beta criteria. I think we should be able to add an example there. I'm less certain about the exact timeline of the Nvidia one. I could remove that one for now and add it back AFTER it's done. wdyt?

Reviewer: let's not take dependencies on consumers we're not sure will be ready as a beta graduation criteria ... one example use seems sufficient

will rollout across nodes.
-->

- kube-controller-manager: If the kube-controller-manager fails to create `ResourceClaim` objects from `ResourceClaimTemplate` due to misconfigurations or permission issues relating to `adminAccess`, then the associated Pods will remain in a pending state and won't be scheduled.

> - kube-scheduler: Bugs in the scheduler might lead to Pods not being scheduled even when resources are available or, scheduling Pods that shouldn't be scheduled due to unmet `adminAccess` requirements. If the `DRAAdminAccess` feature gate isn't enabled or is misconfigured, the scheduler might not recognize ResourceClaim requirements, leading to scheduling failures.

Reviewer: is this thinking of something more than generic scheduler backoff behavior when it encounters failed API requests?

Author: No, this should be part of the generic scheduler backoff behavior.

Reviewer: ok, maybe clarify that... otherwise this line sounds scarier or more specific to this feature than it actually is

logs or events for this purpose.
-->

> ".status.allocation.devices.results[*].adminaccess" will be set to true for a claim using adminAccess when needed by a pod.

Reviewer: ```suggestion
".status.allocation.devices.results[*].adminAccess" will be set to true for a claim using adminAccess when needed by a pod.
```

    - Testing: Are there any tests for failure mode? If not, describe why.
-->

- kube-scheduler cannot allocate ResourceClaims with AdminAccess.

  - Detection: When pods fail to get scheduled, kube-scheduler reports that
    through events and pod status. For DRA, messages include "cannot allocate
    all claims" (insufficient resources) and "ResourceClaim not created yet"
    (user or kube-controller-manager haven't created the ResourceClaim yet).
    The
    ["unschedulable_pods"](https://github.com/kubernetes/kubernetes/blob/9fca4ec44afad4775c877971036b436eef1a1759/pkg/scheduler/metrics/metrics.go#L200-L206)
    metric will have pods counted under the "dynamicresources" plugin label.

    To troubleshoot, "kubectl describe" can be used on (in this order) Pod
    and ResourceClaim.

  - Mitigations: When ResourceClaims or ResourceClaimTemplates the `AdminAccess`

> field don't get created, debugging should focus on the namespace labels. The kube-controller-manager logs should have more information.

Reviewer 2: ```suggestion
field doesn't get created, debugging should focus on the namespace labels. The kube-controller-manager logs should have more information.
```
